tt_eager/tt_dnn/op_library/moreh_norm/moreh_norm_op.cpp

```diff
@@ -0,0 +1,165 @@
+}
+
+std::vector<Shape> MorehNorm::compute_output_shapes(const std::vector<Tensor> &) const { return {}; }
+
+std::vector<Tensor> MorehNorm::create_output_tensors(const std::vector<Tensor> &) const { return {}; }
+
+[[maybe_unused]] Tensor moreh_norm(
+    const Tensor &input,
+    float p,
+    std::optional<std::variant<int64_t, std::vector<int64_t>>> dim,
+    const std::optional<std::reference_wrapper<const Tensor>> output,
+    const MemoryConfig &output_mem_config) {
+    if (dim == std::nullopt) {
+        std::vector<int64_t> dims(input.shape().rank());
+        std::iota(dims.begin(), dims.end(), 0);
+        dim = std::make_optional(dims);
+    }
+
+    if (std::holds_alternative<int64_t>(dim.value())) {
+        const auto d = std::get<int64_t>(dim.value());
+        if (output.has_value()) {
+            return moreh_norm_impl(input, p, d, output->get());
+        }
+        return moreh_norm_impl(input, p, d, create_output_tensor(input, d, output_mem_config));
+    }
+
+    auto dims = std::get<std::vector<int64_t>>(dim.value());
+
+    if (dims.size() == 1) {
+        const auto d = dims[0];
+        if (output.has_value()) {
+            return moreh_norm_impl(input, p, d, output->get());
+        }
+        return moreh_norm_impl(input, p, d, create_output_tensor(input, d, output_mem_config));
+    }
+
+    std::sort(dims.begin(), dims.end(), std::greater<int64_t>());
+    const auto innermost_dim = dims[0];
+    const auto outermost_dim = dims[dims.size() - 1];
+
+    auto tmp_output =
+        moreh_norm_impl(input, p, innermost_dim, create_output_tensor(input, innermost_dim, output_mem_config));
+
+    using idx_t = decltype(dims.size());
+    for (idx_t idx = 1; idx < dims.size() - 1; ++idx) {
+        tmp_output =
+            moreh_norm_impl(tmp_output, p, dims[idx], create_output_tensor(tmp_output, dims[idx], output_mem_config));
+    }
+
+    if (output.has_value()) {
+        return moreh_norm_impl(tmp_output, p, outermost_dim, output->get());
+    }
+    return moreh_norm_impl(
+        tmp_output, p, outermost_dim, create_output_tensor(tmp_output, outermost_dim, output_mem_config));
+}
+
+Tensor moreh_norm_impl(const Tensor &input, float p, int64_t dim, const Tensor &output) {
+    operation::run(MorehNorm{.p = p, .dim = dim}, {input, output});
+    return output;
+}
+
+operation::ProgramWithCallbacks MorehNorm::create_program(
+    const std::vector<Tensor> &input_tensors, std::vector<Tensor> &output_tensors) const {
+    const auto &input = input_tensors.at(0);
+    const auto &output = input_tensors.at(1);
+
+    const auto dim = this->dim;
+    const auto input_rank = static_cast<decltype(dim)>(input.shape().rank());
+
+    if (dim == input_rank - 1) {
+        return moreh_norm_w_impl(input, this->p, output);
+    } else if (dim == input_rank - 2) {
+        return moreh_norm_h_impl(input, this->p, output);
+    } else {
+        return moreh_norm_other_impl(input, this->p, dim, output);
```

Comment:
> It's difficult to understand what 'other_impl' mean.

I agree that 'other_impl' is not a good name.
The reason for not using 'norm_nc' is because if the default dimension of the Tensor changes(4D -> 5D) in the future, it may be changed to something like 'ncd' instead of 'nc'.
Additionally, if it is changed to 'nc' now, there are too many places(folder, file, code...) that need to be fixed.